### ISSUE TITLE
Respect query constraints within relations when batchloading relationships

### ISSUE THREAD

**Issue by kajdehoop** [nuwave/lighthouse#815]
**Describe the bug**

I have a @hasOne relationship on one of my models, and I want to return this relationship in GraphQL, but the result is always null.

**Expected behavior**
I expect the selected_guide_booking field to return a model according to the @hasOne relationship.

In my other Laravel code (the blade files for example) this works perfectly. `$booking->selected_guide_booking->price` for example will return the right price. However when I do the same in a GraphQL query, it always returns null.

**Schema**

This is the relationship in my model:
```
public function selected_guide_booking()
    {
        return $this->hasOne('App\GuideBooking','booking_id')->where('guide_id', $this->guide_id);
    }
```
This is the field on my Booking Type:
```
selected_guide_booking: GuideBooking @hasOne
```
Each Booking can have multiple GuideBookings, but only has one GuideBooking where the guide_id is the same as the guide_id in the Booking.

**Output/Logs**

<details><summary>Example GraphQL response, click to expand</summary>

```
{
  "data": {
    "me": {
      "email": "test@test.com",
      "price": "35",
      "completed_incomingbookings": [
        {
          "selected_guide_booking": null,
          "review": {
            "id": "1",
            "text": "Bla bla bla bla",
            "ranking": "5",
            "tourist": {
              "realname": "Madeline Hayes MD",
              "photo": null,
              "about_tourist": "Adipisci magnam corrupti officia nulla. Sunt cupiditate nam consequatur sed omnis. Sit non voluptatem voluptatem autem. Repudiandae eum vero quia reiciendis qui nam. Quia ut ut quam. Alias reiciendis est vitae sint sint et."
            }
          }
        }
      ]
    }
  }
}
```
The `review` field on this query also uses a @hasOne relationship, but does work correctly. As far as I can tell the only difference is that the selected_guide_booking in theory could return multiple models, if it wouldn't have the 'where' clause on the relationship.
</details></br>

**Environment**

Lighthouse Version: 2.6.4
Laravel Version: 5.7.26
PHP Version: 7.2.11

**Additional context**

I've tried changing the @hasOne relationship on the selected_guide_booking field to a @hasMany relationship. This correctly returns the model, but it is inside an array. This breaks the other Laravel code, because it's expecting a hasOne relationship. A temporary fix could be to use a hasMany relationship and then putting `[0]` behind every returned model to retrieve the first model in the array, but this would make for some ugly code.

**Comment by spawnia** [comment]
@kajdehoop thanks for reporting this issue.

When you use relationship directives such as `@hasOne`, Lighthouse expects the underlying method to actually return a relation class, in this case `\Illuminate\Database\Eloquent\Relations\HasOne`.

The method `selected_guide_booking` in your model is not *actually* a relationship in that sense. It is rather linked to another model semantically, using the actual underyling relation as a basis.

I think what you can do in this case is simply to omit the `@hasOne` directive. You will not get the batched query optimization that Lighthouse provides you with, so you may experience a slight performance degregation for some queries.

I noticed you are still on `2.6.4`, can you try upgrading to the latest version and see if that changes anything?

**Comment by kajdehoop** [comment]
@spawnia Thank you very much! Removing the `@hasOne` directive fixed the issue. Have a great day (:

**Comment by palypster** [comment]
@spawnia is this really a final solution? I have similar problem with `hasMany` directive. My schema looks like this:
```
type StructuredBudgetItem {
    id: ID!
    budgetItems: [BudgetItem!] @hasMany
}
```
and my model:
```
class StructuredBudgetItem extends Model
{
    public function budgetItems(): HasMany
    {
        return $this->hasMany(BudgetItem::class, 'budget_category_id')
            ->where('project_id', $this->project_id);
    }
}
```
If I query `budgetItems`, I get empty array. If I remove `hasMany` directive, everything start to work, but I'm missing batch/eager loading (so I get tons of queries).

Can you suggest how can this be solved/workarounded in a proper way?

**Comment by palypster** [comment]
I probably don't understand your explanation, because it doesn't work even if I do this:
```
public function budgetItems(): HasMany
    {
        $relation = $this->hasMany(BudgetItem::class, 'budget_category_id');

        $relation->where('project_id', $this->project_id);

        return $relation;
    }
```
In this case it really returns relation instance, so why it doesn't work? Could you explain it a bit more? Thanks!

**Comment by lorado** [comment]
I mean `@hasMany` and `@hasOne` use their own relation resolving (I am not 100% sure, but I saw something like this some time ago...), so it is possible to implement batch loading. In my code I have some complex logic, but at the end I still return `HasMany`/ `HasOne` object, and my customization is simply ignored, if I use those directives. I currently accept the lack of batch loading, but so I get correct data and for currently amount of data it is not a big performance problem. I would like to improve those directives, but I currently have no time for it =\

**Comment by palypster** [comment]
@lorado okay, I think I understand. But if `@hasMany` and `@hasOne` use their own relation resolving, that would mean I don't need to create the relation in the model at all, right? But when I try that:
```
class StructuredBudgetItem extends Model
{
    
}
```
I get error: `Call to undefined relationship [budgetItems] on model [App\\StructuredBudgetItem].`

So it looks the defined relation is taken into an account, but not fully as is, but get somehow modificated maybe. This is really weird and unexpected.

I understand the lack of time, maybe we can at least reopen this issue so maybe in the future somebody have the time.. ;)

**Comment by lorado** [comment]
Yes I think you still has to define the relation, but it take only some part of it or kinda... Let's wait, what @spawnia says...

**Comment by spawnia** [comment]
I suspect this might already be resolved for simple relations in the latest https://github.com/nuwave/lighthouse/releases/tag/v4.7.2, can you try?

The responsible code can be found in https://github.com/nuwave/lighthouse/blob/master/src/Execution/DataLoader/RelationBatchLoader.php

However, the behaviour you are experiencing might still be present when batchloading paginated relations, see https://github.com/nuwave/lighthouse/blob/master/src/Execution/DataLoader/ModelRelationFetcher.php

**Comment by palypster** [comment]
I have updated to v4.7.2 and it is still not working. It behaves exactly as before (in v4.7.1) - if I use `@hasMany` directive, I get empty array. In the SQL log I can see:
```...where  "project_id" is null...``` so the problem is that $this->project_id is probably null or something.

Note that if I try to load the relation using i.e. `artisan tinker`, everything works even with eager loading.

**Comment by SamuelT-ST** [comment]
I have same problem. 

When I use where on relationship in model, relation in GraphQl returns empty array. I can use it without @belongsTo but than there is too many queries (eager loading does not work).

**Comment by palypster** [comment]
@spawnia any chance this being fixed soon also for @paginated results?

**Comment by spawnia** [comment]
@palypster feel free to provide a PR. You do not have to solve this right away, a failing test case is a good starting point.
